STYLEGUIDE.md

```diff
@@ -3,25 +3,695 @@
+
+### Prefer modern JavaScript/TypeScript syntax
+
+You should prefer modern language features in a lot of cases, e.g.:
+
+* Prefer `class` over `prototype` inheritance
+* Prefer template strings over string concatenation
+
+### Avoid mutability and state
+
+Wherever possible, do not rely on mutable state. This means you should not
+reassign variables, modify object properties, or push values to arrays.
+Instead, create new variables, and shallow copies of objects and arrays:
+
+```js
+// good
+function addBar(foos, foo) {
+  const newFoo = {...foo, name: 'bar'};
+  return [...foos, newFoo];
+}
+
+// bad
+function addBar(foos, foo) {
+  foo.name = 'bar';
+  foos.push(foo);
+}
+```
+
+### Return/throw early from functions
+
+To avoid deep nesting of if-statements, always return a function's value as early
+as possible. And where possible, do any assertions first:
+
+```js
+// good
+function doStuff(val) {
+  if (val > 100) {
+    throw new Error('Too big');
+  }
+
+  if (val < 0) {
+    return false;
+  }
+
+  // ... stuff
+}
+
+// bad
+function doStuff(val) {
+  if (val >= 0) {
+    if (val < 100) {
+      // ... stuff
+    } else {
+      throw new Error('Too big');
+    }
+  } else {
+    return false;
+  }
+}
+```
+
+### Use object destructuring
+
+This helps avoid temporary references and helps prevent typo-related bugs.
+
+```js
+// best
+function fullName({ first, last }) {
+  return `${first} ${last}`;
+}
+
+// good
+function fullName(user) {
+  const { first, last } = user;
+  return `${first} ${last}`;
+}
+
+// bad
+function fullName(user) {
+  const first = user.first;
+  const last = user.last;
+  return `${first} ${last}`;
+}
+```
+
+### Use array destructuring
+
+Directly accessing array values via index should be avoided, but if it is
+necessary, use array destructuring:
+
+```js
+const arr = [1, 2, 3];
+
+// good
+const [first, second] = arr;
+
+// bad
+const first = arr[0];
+const second = arr[1];
+```
+
+### Declare one variable per line, wherever it makes the most sense
+
+This makes it easier to re-order the lines. However, declari variables
+wherever it makes sense (not necessairly on the top of a function).
+
+```js
+// good
+const keys = ['foo', 'bar'];
+const values = [23, 42];
+
+// bad
+const keys = ['foo', 'bar'],
+      values = [23, 42];
+```
+
+### Prefix private class methods with an underscore `JS only`
+
+Identifying private class methods makes it easier to differentiate a class's public and internal
+APIs, and makes private methods easier to mark as `private` when the code is migrated to TypeScript.
+
+```js
+// good
+class BankAccount {
+  addFunds() {}
+
+  _calculateInterest() {}
+}
+```
+
+If using TypeScript you should use the `private` modifier instead.
+
+### Magic numbers/strings
+
+These are numbers (or other values) simply used in line in your code. *Do not
+use these*, give them a variable name so they can be understood and changed
+easily.
+
+```js
+// good
+const minWidth = 300;
```

Comment:
I know we had a separate discussion awhile back about how to name constants but i would do it this way:

```suggestion
const MIN_WIDTH = 300;
```

see https://github.com/elastic/kibana/issues/11288#issuecomment-297565969.  It was closed somewhat unresolved but the last comment had 8 thumb ups for 

>Add a 👍 if you prefer to add CAPS_AND_UNDERSCORES for constants to the style guide, and 👎 if you prefer not to.

Reply:
If it would be fine for you I would also like to revisit this (since it's not just moving the rule, but actually changing something) in a separate PR and will use that voting into account (and also figure out if we can have a linting rule for that)

Reply:
sgtm!